core/src/main/scala/akka/persistence/jdbc/state/scaladsl/JdbcDurableStateStore.scala

```diff
@@ -0,0 +1,209 @@
+
+  def changes(tag: String, offset: Offset): Source[DurableStateChange[A], NotUsed] =
+    _changesByTag(tag, offset.value, terminateAfterOffset = None)
+
+  private def currentChangesByTag(tag: String, offset: Long): Source[DurableStateChange[A], NotUsed] = {
+    Source
+      .futureSource(maxJournalSequence().map { maxOrderingInDb =>
+        _changesByTag(tag, offset, terminateAfterOffset = Some(maxOrderingInDb))
+      })
+      .mapMaterializedValue(_ => NotUsed)
+  }
+
+  private def _currentChangesByTag(
+      tag: String,
+      from: Long,
+      batchSize: Long,
+      queryUntil: MaxOffset): Source[DurableStateChange[A], NotUsed] = {
+    if (queryUntil.value < from.value) Source.empty
+    else changesByTag(tag, from, queryUntil.value, batchSize).mapAsync(1)(Future.fromTry)
+  }
+
+  private def changesByTag(
+      tag: String,
+      offset: Long,
+      maxOffset: Long,
+      max: Long): Source[Try[DurableStateChange[A]], NotUsed] = {
+    Source
+      .fromPublisher(db.stream(queries.changesByTag((tag, offset, maxOffset, max)).result))
+      .map(toDurableStateChange)
+  }
+
+  // checks if the last n elements of a list are equal
+  private def equalLastNElements(coll: List[Long], n: Int): Boolean = {
+    if (coll.isEmpty) false
+    else if (coll.size < n) false
+    else {
+      val res = coll.slice(coll.length - n, coll.length)
+      res.forall(_ == res.head)
+    }
+  }
+
+  private def _changesByTag(
+      tag: String,
+      offset: Long,
+      terminateAfterOffset: Option[Long]): Source[DurableStateChange[A], NotUsed] = {
+    val batchSize = durableStateConfig.batchSize
+    val startingOffsets = collection.mutable.ListBuffer.empty[Long]
+    val stopAfterEmptyFetchIterations = durableStateConfig.stopAfterEmptyFetchIterations
+    Source
+      .unfoldAsync[(Long, FlowControl), Seq[DurableStateChange[A]]]((offset, Continue)) { case (from, control) =>
+        def retrieveNextBatch() = {
+          for {
+            queryUntil <- maxJournalSequence()
+            xs <- _currentChangesByTag(tag, from, batchSize, MaxOffset(queryUntil)).runWith(Sink.seq)
+          } yield {
+            val hasMoreEvents = xs.size == batchSize
+            val nextControl: FlowControl =
+              terminateAfterOffset match {
+                // we may stop if target is behind queryUntil and we don't have more events to fetch
+                case Some(target) if !hasMoreEvents && target <= queryUntil => Stop
+
+                // We may also stop if we have found an event with an offset >= target
+                case Some(target) if xs.exists(_.offset.value >= target) => Stop
+
+                // otherwise, disregarding if Some or None, we must decide how to continue
+                case _ =>
+                  if (
+                    stopAfterEmptyFetchIterations > 0 && equalLastNElements(
+                      startingOffsets.toList,
+                      stopAfterEmptyFetchIterations)
```

Comment:
this case is only intended for tests, wonder if those tests could be written in another way to avoid this additional stuff and the mutable `startingOffsets`

Reply:
@patriknw The problem is the `source` in this case is live and hence when I do a `source.runWith(Sink.ignore)` I get a `Future` which does not complete. The API `changes(persistenceId, tag)` does not give me any handle to complete the future. And also future cancelation is a hairy task. Hence for tests, I had to resort to such tactics for completing the future. Any suggestions welcome ..

Reply:
You could perhaps use `takeWhile` or `take` to limit downstream in the test instead?

Reply:
@johanandren Even if I make `takeWhile` or `take` where would I do that ? All queries are in the main code and not in tests. The only API I have in tests in `changes(persistencId, tag)`. Hence my logic was if I get a preset number of 0 results in query I stop.

Reply:
@johanandren Now fixed. Using `takeWhile` .. Thanks.